lib/stitch_plan/lock_stitch.py

```diff
@@ -0,0 +1,202 @@
+        for stitch in stitches[1:]:
+            to_start = stitch - upcoming[-1]
+            upcoming.append(stitch)
+            if to_start.length() >= max_lock_length:
+                break
+        line = LineString(upcoming)
+
+        # add tie stitches
+        lock_stitches = []
+        for i, tie_path in enumerate(lock_pos):
+            if tie_path < 0:
+                stitch = Stitch(stitches[0] + tie_path * (stitches[1] - stitches[0]).unit())
+            else:
+                stitch = substring(line, start_dist=tie_path, end_dist=tie_path)
+                stitch = Stitch(stitch.x, stitch.y, tags=('lock_stitch',))
+            lock_stitches.append(stitch)
+        return lock_stitches
+
+    def _relative_tie(self, stitches, pos):
+        path = string_to_floats(self.path, " ")
+
+        to_previous = stitches[1] - stitches[0]
+        length = to_previous.length()
+
+        lock_stitches = []
+        if length > 0.5 * PIXELS_PER_MM:
+
+            # Travel back one stitch, stopping halfway there.
+            # Then go forward one stitch, stopping halfway between
+            # again.
+
+            # but travel at most 1.5 mm
+            length = min(length, 1.5 * PIXELS_PER_MM)
+
+            direction = to_previous.unit()
+
+            for delta in path:
+                lock_stitches.append(Stitch(stitches[0] + delta * length * direction, tags=('lock_stitch')))
+        else:
+            # Too short to travel part of the way to the previous stitch; just go
+            # back and forth to it a couple times.
+            for i in (1, 0, 1, 0):
+                lock_stitches.append(stitches[i])
+        return lock_stitches
+
+
+def get_lock_stitch_by_id(pos, lock_type, default="half_stitch"):
+    id_list = [str(lock) for lock in LOCK_DEFAULTS[pos]]
+
+    lock = LOCK_DEFAULTS[pos][id_list.index(lock_type)] or None
+    if lock is None:
+        lock = LOCK_DEFAULTS[pos][id_list.index(default)]
+    return lock
+
+
+half_stitch = LockType("half_stitch", _("Half Stitch"), "0 0.5 1 0.5 0", "relative")
+arrow = LockType("arrow", _("Arrow"), "M 0.5,0.3 0.3,1.31 -0.11,0.68 H 0.9 L 0.5,1.31 0.4,0.31 V 0.31 1.3", "svg")
+back_forth = LockType("back_forth", _("Back and forth"), "1 1 -1 -1", "absolute")
+bowtie = LockType("bowtie", _("Bowtie"), "M 0,0 -0.39,0.97 0.3,0.03 0.14,1.02 0,0 V 0.15", "svg")
+cross = LockType("cross", _("Cross"), "M 0,0 -0.7,-0.7 0.7,0.7 0,0 -0.7,0.7 0.7,-0.7 0,0 -0,-0.7", "svg")
+star = LockType("star", _("Star"), "M 0.67,-0.2 C 0.27,-0.06 -0.22,0.11 -0.67,0.27 L 0.57,0.33 -0.5,-0.27 0,0.67 V 0 -0.5", "svg")
+simple = LockType("simple", _("Simple"), "M -0.03,0 0.09,0.81 0,1.49 V 0 0.48", "svg")
+triangle = LockType("triangle", _("Triangle"), "M -0.26,0.33 H 0.55 L 0,0.84 V 0 L 0.34,0.82", "svg")
+zigzag = LockType("zigzag", _("Zig-zag"), "M -0.25,0.2 0.17,0.77 -0.22,1.45 0.21,2.05 -0.03,3 0,0", "svg")
+custom = LockType("custom", _("Custom"), path_type="absolute svg")
```

Comment:
Why is this "absolute svg"?  I think it should be just "svg"?

Reply:
No, custom is actually both.